weco/optimizer.py

```
@@ -543,8 +900,794 @@ def signal_handler(signum, frame):
+    run_log_dir = pathlib.Path(log_dir) / run_id
+    run_log_dir.mkdir(parents=True, exist_ok=True)
+
+    # Determine the source file path
+    if source_path_from_api and pathlib.Path(source_path_from_api).exists():
+        source_path = source_path_from_api
+        console.print(f"[cyan]Using source file from original run: {source_path}[/]")
+    else:
+        # Try to find the source file
+        possible_files = []
+        for pattern in ["train.py", "main.py", "solution.py", "*.py"]:
+            files = list(pathlib.Path(".").glob(pattern))
+            possible_files.extend(files)
+
+        possible_files = list(set(f for f in possible_files if f.is_file()))
+
+        if len(possible_files) == 1:
+            source_path = str(possible_files[0])
+            console.print(f"[cyan]Found source file: {source_path}[/]")
+        else:
+            console.print("[yellow]Please specify the source file to continue optimizing.[/]")
+            source_path = console.input("[bold]Enter the path to the source file: [/]").strip()
+            if not pathlib.Path(source_path).exists():
+                console.print(f"[bold red]Source file not found: {source_path}[/]")
+                return False
+
+    # Write the last completed solution (starting point) to the source file
+    # We continue from where we left off (last completed step)
+    if last_node and last_node.get("code"):
+        write_to_path(pathlib.Path(source_path), last_node["code"])
+        write_to_path(run_log_dir / f"step_{last_step}.py", last_node["code"])
+        console.print(
+            f"\n[green]Extending from last completed step {last_step} (metric: {last_node.get('metric_value', 'N/A')}):[/] {source_path}"
+        )
+    else:
+        # Fallback to original source code if no solution available
+        write_to_path(pathlib.Path(source_path), source_code)
+        console.print(f"\n[yellow]No last solution found, starting from original source:[/] {source_path}")
```

I don't we should have a fallback here. If the last_node doesn't exist, we should raise an error to the user.